File: internal/services/container/nodekind/specs/expr.go
// internal/services/container/nodekind/specs/expr.go
package specs

import "yhbox/internal/services/container/nodekind"

func init() {
	nodekind.Register(&nodekind.Spec{
		Kind:    "Expr",
		Group:   "variables",
		ExecIn:  nil,
		ExecOut: nil,
		DataOut: map[string]nodekind.PinType{"value": nodekind.PinAny},
		// Expr 的 data-in pin 由 config.inputs[] 动态生成.
		// 每个 input = {name: string, type: PinType}. 校验/eval/前端渲染全走 DataInDynamicFn.
		DataInDynamicFn: func(cfg map[string]any) map[string]nodekind.PinType {
			out := map[string]nodekind.PinType{}
			inputs, _ := cfg["inputs"].([]any)
			for _, raw := range inputs {
				obj, _ := raw.(map[string]any)
				name, _ := obj["name"].(string)
				typ, _ := obj["type"].(string)
				if name == "" || typ == "" {
					continue
				}
				out[name] = nodekind.PinType(typ)
			}
			return out
		},
		Defaults:   map[string]any{"expr": "", "outType": "auto", "inputs": []any{}},
		IsPureData: true,
	})
}

File: frontend/src/components/containers/nodeRegistry/specs/expr.ts
// frontend/src/components/containers/nodeRegistry/specs/expr.ts
// Expr (1) — dynamic data-in pins derived from config.inputs[].
// Mirrors backend internal/services/container/nodekind/specs/expr.go.
import { register } from '../registry'
import type { PinType } from '../index'

register({
  kind: 'Expr',
  group: 'variables',
  labelZh: '表达式',
  description: '',
  visual: { icon: 'i-tabler-math', bg: 'bg-amber-500/15', border: 'border-amber-500/40' },
  execIn: [],
  execOut: [],
  dataIn: {},
  dataInDynamicFn: (cfg) => {
    const out: Record<string, PinType> = {}
    const inputs = Array.isArray(cfg?.inputs) ? (cfg!.inputs as unknown[]) : []
    for (const raw of inputs) {
      const obj = raw as Record<string, unknown> | null
      const name = typeof obj?.name === 'string' ? obj.name : ''
      const type = typeof obj?.type === 'string' ? (obj.type as PinType) : ('' as PinType | '')
      if (!name || !type) continue
      out[name] = type as PinType
    }
    return out
  },
  dataOut: { value: 'any' },
  fields: [
    {
      key: 'expr',
      label: '表达式',
      type: 'text',
      placeholder: 'i + 1',
    },
    {
      key: 'outType',
      label: '输出类型',
      type: 'select',
      options: [
        { value: 'auto', label: 'auto (默认, 推断)' },
        { value: 'number', label: 'number' },
        { value: 'bool', label: 'bool' },
        { value: 'string', label: 'string' },
        { value: 'point', label: 'point' },
        { value: 'any', label: 'any' },
      ],
    },
  ],
  defaults: { expr: '', outType: 'auto', inputs: [] },
  isPureData: true,
})
